File: algorithms/cpp/previousPermutationWithOneSwap/previousPermutationWithOneSwap.cpp
// Source: https://leetcode.com/problems/previous-permutation-with-one-swap/
// Author: Miao Zhang
// Date:   2021-04-07

class Solution {
public:
    vector<int> prevPermOpt1(vector<int>& arr) {
        int n = arr.size();
        int maxval = 0;
        int idx = -1;
        for (int i = n - 1; i > 0; i--) {
            if (arr[i] >= arr[i - 1]) continue;
            for (int j = i; j < n; j++) {
                if (arr[j] < arr[i - 1] && maxval < arr[j]) {
                    maxval = arr[j];
                    idx = j;
                }
            }
            swap(arr[i- 1], arr[idx]);
            break;
        }
        return arr;
    }
};

File: algorithms/python/previousPermutationWithOneSwap/previousPermutationWithOneSwap.py
#! /usr/bin/env python3
# -*- coding: utf-8 -*-

# Source: https://leetcode.com/problems/previous-permutation-with-one-swap/
# Author: Miao Zhang
# Date:   2021-04-07

class Solution:
    def prevPermOpt1(self, arr: List[int]) -> List[int]:
        n = len(arr)
        idx = -1
        maxval = 0
        for i in range(n - 1, 0, -1):
            if arr[i] >= arr[i - 1]: continue
            for j in range(i, n):
                if arr[j] < arr[i - 1] and maxval < arr[j]:
                    maxval = arr[j]
                    idx = j
            arr[i - 1], arr[idx] = arr[idx], arr[i - 1]
            break
        return arr
